### ISSUE TITLE
[Experiment] Joint AOPE Recall Boost: AfroXLMR + Explicit Filtering + BIO Weighting (Run 1)

### ISSUE THREAD

**Issue by AzariyasMekonen** [EthAIResearch/acos-amharic#58]
## Experiment Overview

- **Model Run**: `results/stage_aope_sdrn/afroxlmr_base_run1`
- **Architecture**: Synchronous Double-channel Recurrent Network (`JointAOPESDRN`)
- **Backbone**: `Davlan/afro-xlmr-base`
- **Training Data**: `data/prepared_explicit/train.jsonl` (explicit quad subset, removing zero-label background dilution)
- **Evaluation Data**: `data/prepared_explicit/test.jsonl` (2,963 gold explicit pairs)
- **Loss Parameters**:
  - `bio_class_weights`: `[1.0, 3.0, 3.0]` (upweights `B` and `I` 3x over `O`)
  - `span_loss_weight`: `2.0`
  - Recurrent steps: `2`

---

## Motivation & Prior Bottleneck

In `roberta_base_run1`, AOPE pair recall plateaued at **24.23%**, with an empirical candidate ceiling of **24.87%**. Offline threshold sweeping revealed that the bilinear relation matrix was not rejecting pairs prematurely, but rather that the span extraction heads suffered from low recall (Aspect: 44.22%, Opinion: 34.45%) due to the ~42% zero-explicit sentence noise in `data/prepared/`.

This experiment tested the hypothesis that filtering to explicit sentences, applying class-weighted CrossEntropy loss, and switching to the African multilingual backbone `Davlan/afro-xlmr-base` would shatter this span recall ceiling.

---

### Comparison: `roberta_base_run1` vs `afroxlmr_base_run1`

| Metric | Baseline (`roberta_base_run1`) | New Run (`afroxlmr_base_run1`) | Net Delta |
| :--- | :--- | :--- | :--- |
| **Dataset** | `data/prepared/` | `data/prepared_explicit/` | Zero-label noise filtered |
| **Backbone** | `rasyosef/roberta-base-amharic` | `Davlan/afro-xlmr-base` | African Multilingual LM |
| **Loss Weights** | Unweighted | `bio: [1.0, 3.0, 3.0]`, $\lambda_{\text{span}}=2.0$ | Class rebalanced |
| **Aspect Span Recall** | 44.22% (1,297 TP) | **63.79% (1,871 TP)** | **+19.57% (+574 TP)** |
| **Aspect Span Precision** | 48.14% | **53.89%** | **+5.75%** |
| **Aspect Span F1** | 46.10% | **58.42%** | **+12.32%** |
| **Opinion Span Recall** | 34.45% (1,016 TP) | **48.32% (1,425 TP)** | **+13.87% (+409 TP)** |
| **Opinion Span Precision** | 37.11% | **40.49%** | **+3.38%** |
| **Opinion Span F1** | 35.73% | **44.06%** | **+8.33%** |
| **Theoretical Pair Recall Ceiling** | 15.24% | **30.82%** | **+15.58%** |
| **Empirical Candidate Ceiling** | 24.87% (737 TP) | **35.88% (1,063 TP)** | **+11.01% (+326 TP)** |
| **Pair Precision ($\hat{\delta}=0.5$)** | 28.99% | **35.31%** | **+6.32%** |
| **Pair Recall ($\hat{\delta}=0.5$)** | 24.20% (717 TP) | **30.54% (905 TP)** | **+6.34% (+188 TP)** |
| **Pair F1 ($\hat{\delta}=0.5$)** | 26.38% | **32.75%** | **+6.37%** |
| **Peak F1 ($\hat{\delta}=0.55$)** | 26.43% | **32.82%** | **+6.39%** |

---

## Threshold Sweep Analysis

Full threshold sweep on `results/stage_aope_sdrn/afroxlmr_base_run1/threshold_sweep.json`:

```
Candidate Ceiling: [1,063 TP / 3,230 FP] -> Recall = 35.88%, Precision = 24.76%

Threshold (δ̂) |  Precision |  Recall   |    F1     |   TP   |   FP   |   FN
-------------------------------------------------------------------------
    0.05      |   25.13%   |  35.74%   |  29.51%   |  1,059 |  3,155 | 1,904
    0.10      |   25.84%   |  35.40%   |  29.88%   |  1,049 |  3,010 | 1,914
    0.15      |   26.67%   |  35.07%   |  30.30%   |  1,039 |  2,857 | 1,924
    0.20      |   27.61%   |  34.63%   |  30.72%   |  1,026 |  2,690 | 1,937
    0.25      |   28.46%   |  34.02%   |  30.99%   |  1,008 |  2,534 | 1,955
    0.30      |   29.65%   |  33.45%   |  31.44%   |    991 |  2,351 | 1,972
    0.35      |   30.94%   |  32.87%   |  31.88%   |    974 |  2,174 | 1,989
    0.40      |   32.26%   |  32.40%   |  32.33%   |    960 |  2,016 | 2,003
    0.45      |   33.49%   |  31.59%   |  32.51%   |    936 |  1,859 | 2,027
    0.50      |   35.31%   |  30.54%   |  32.75%   |    905 |  1,658 | 2,058
    0.55 (★)  |   36.78%   |  29.63%   |  32.82%   |    878 |  1,509 | 2,085
    0.60      |   38.85%   |  28.38%   |  32.80%   |    841 |  1,324 | 2,122
```

### Key Findings:
1. **Ceiling Smashed**: Candidate ceiling jumped from **24.87% to 35.88%**, confirming that negative sentence dilution and class imbalance were indeed suppressing span recognition.
2. **Aspect Recall Strong at 63.79%**: Up from 44.22%. Aspect extraction is now robust.
3. **Opinion Extraction is the New Frontier**: At 48.32% recall, opinions remain harder to capture than aspects due to morphological variance and negation compounding in Amharic.
4. **Optimal Threshold Recommendations**:
   - For standalone pair extraction: $\hat{\delta} = 0.55$ yields peak F1 of **32.82%**.
   - For feeding downstream quad pipeline (Stages 3 & 4): $\hat{\delta} = 0.25 - 0.35$ provides higher recall (**33.5% - 34.0%**, ~1,000 true pairs) with good precision (~29%).

---

## Artifacts & Links
- Results directory: `results/stage_aope_sdrn/afroxlmr_base_run1/`
- Pull Request: #56
- Related Issue: #55
